### PR TITLE
Make the plugin LSP 1.1 compatible

### PR DESCRIPTION
**Pull request by rwols**
I have a draft pull request where I remove the LanguageConfig entirely and replace it by a selector. This is much more practical. See: https://github.com/sublimelsp/LSP/pull/1428

Because there are many other improvements for the ST4 version I decided to add all improvements as well. For instance it's fairly easy to write custom notifications and request handlers with the AbstractPlugin class. It's also trivial now to put a status message on all views attached to the session. And also registration of plugins is now explicit (in plugin_loaded, plugin_unloaded) which allows users to put the package in their `"ignored_packages"` setting and then LSP will pick up this change in state and stop the server subprocess (likewise for when the user enables the package).

I also moved the helper py files to their own directory to ensure that ST doesn't load these files as separate plugins (but merely, they are imported by a single plugin file called plugin.py).

The changes ensure compatibility with both ST3 and ST4. Though I haven't tested if everything still works on ST3 with these changes.

Don't merge/release this unless the above mentioned pull request is merged.

### CODE REVIEW

**Comment by ayoub-benali** [comment]
@rwols thanks for taking care of this. I will review it later today :)

**Comment by rchl** [comment]
I'm working on lsp_utils refactoring that would hopefully make it easier to rewrite this in a more concise matter.
I should have it ready in a few days or so.

(Do note I haven't checked the requirements of this plugin so I'm not sure how suitable that will be).

File: LSP-metals.sublime-settings [hunk 1]
```diff
@@ -13,6 +20,30 @@
     "-Xms100m"
   ],
 
-  // Optional path to the Java home directory. Requires reloading sublime.
-  "java_home": ""
+  // The startup command (ST4 only)
+  "command": [
```
Comment (ayoub-benali) [inline LSP-metals.sublime-settings:24]: this shouldn't leak to the user setting. One would care only about `server_properties`

File: LSP-metals.sublime-settings [hunk 2]
```diff
@@ -2,9 +2,16 @@
...
+  // Optional path to the Java home directory. Requires reloading sublime.
+  "java_home": "",
+
+  // Attach to views with base scope source.scala (ST4 only)
+  "selector": "source.scala",
+
   // Optional array of properties to pass along to the Metals server.
   // Example: `-Dhttps.proxyHost=… -Dhttps.proxyPort=…` or `-Dmetals.statistics=all`
   // check https://scalameta.org/metals/docs/editors/new-editor.html#metals-server-properties for all properties
+  // (ST3 only)
```
Comment (ayoub-benali) [inline LSP-metals.sublime-settings:14]: what is blocking from using this in ST4 as well ?

File: LSP-metals.sublime-settings [hunk 3]
```diff
@@ -13,6 +20,30 @@
...
+    "bintray:scalacenter/releases",
+    "--repository",
+    "sonatype:snapshots",
+    "--main-class",
+    "scala.meta.metals.Main",
+    "org.scalameta:metals_2.12:${server_version}"
+  ],
+
+  // initialization options (ST4 only)
+  "initializationOptions": {
```
Comment (ayoub-benali) [inline LSP-metals.sublime-settings:46]: in this PR it is used only for ST3, `initializationOptions` should be set in both versions

File: st3/__init__.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+                scopes=["source.scala"],
+                syntaxes=["Packages/Scala/Scala.sublime-syntax"])
+
+        metals_config = ClientConfig(
+            name=server_name,
+            binary_args=launch_command,
+            tcp_port=None,
+            languages=[language],
+            enabled=True,
+            init_options=DottedDict(plugin_settings.get("initializationOptions")),
```
Comment (rwols) [inline st3/__init__.py:46]: I added DottedDict. I'm not sure how this previously even worked without DottedDict.

File: st3/__init__.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+            # LSP 1.0
+            language = LanguageConfig("scala")
+        except TypeError:
+            # Ancient LSP
+            language = LanguageConfig(
+                language_id="scala",
+                scopes=["source.scala"],
+                syntaxes=["Packages/Scala/Scala.sublime-syntax"])
+
+        metals_config = ClientConfig(
```
Comment (rchl) [inline st3/__init__.py:40]: Is there a good reason for it not using `from LSP.plugin import read_client_config` ?

File: plugin.py [hunk 6]
```diff
@@ -1,158 +1,24 @@
...
-    none_empty = list(filter(None, stripped))
-    return none_empty
+    sublime.error_message(" ".join(fmt).format(name))
+else:
+    try:
+        from LSP.plugin import __version__ as v
+        if v < (1, 1, 0):
+            from .st3 import *
+        elif (1, 1, 0) <= v < (1, 2, 0):
+            from .st4 import *
```
Comment (rchl) [inline plugin.py:18]: by rules of semantic versioning, the "st4" API should work until 2.0.0, not until 1.2.0.

File: st3/__init__.py (hunk 5 above)
Comment (rwols) [reply st3/__init__.py:40]: There is not :) Might as well change that now that I'm busy with this.

File: st4/__init__.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        return cls.__name__.lower()
+
+    @classmethod
+    def configuration(cls) -> Tuple[sublime.Settings, str]:
+        settings, path = super().configuration()
+        java_path = get_java_path(settings)
+        version = settings.get("server_version")
+        properties = prepare_server_properties(settings.get("server_properties"))
+        command = create_launch_command(java_path, version, properties)
+        settings.set("command", command)
```
[the remark this answers is not in the document]
Comment (rwols) [reply st4/__init__.py:31]: Run "LSP: Dump Window Configs" from the command palette and you can see that the init options are set. Such is the magic of AbstractPlugin ;) no, it does that in the base class.

**Comment by rwols** [comment]
@ayoub-benali did you find the time to test this? Can we merge this soonish?

File: st4/__init__.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ayoub-benali) [reply st4/__init__.py:31]: `LSP: Dump Window Config` is pretty cool ! :)

**Review by ayoub-benali — APPROVED** [review]
@rwols sorry for the delay. I just tested and everything seems to be working fine. Thank you for tacking care of this :)

If I understood correctly once sublimelsp/LSP#1428 is merged I should merge this make a release ?

**Comment by rwols** [comment]
It’s the other way around. These changes make sure that this plugin will work with the linked changes in the LSP repo. So this should be merged and released first.

**Comment by rwols** [comment]
I notice now that I said in my initial comment “don’t merge this until the above mentioned pull request is merged” but that is wrong.

**Comment by ayoub-benali** [comment]
@rwols new [release](https://github.com/scalameta/metals-sublime/releases/tag/0.6.0) is out. Hopefully this would unblock https://github.com/sublimelsp/LSP/pull/1428 and thanks again :)

**Comment by rwols** [comment]
Awesome, thanks!

**Pull request merged**